clients/cppsal/include/sal_node.h

```diff
@@ -0,0 +1,658 @@
+                j->set("_version", version);
+                return j;
+            }
+            */
+
+            static NodeInfo decode(Poco::JSON::Object::Ptr j)
+            {
+                std::string _name;
+                if (j->has("name"))
+                    _name = j->getValue<std::string>("name");
+                NodeInfo i{j->getValue<std::string>("class"), j->getValue<std::string>("group"),
+                           j->getValue<uint64_t>("version"), _name};
+                return i;
+            }
+        };
+
+        ///  map to std::filesystem::perms or  OPC-UA AccessLevel, UserAccessLevel
+        /// RolePermissions
+        /// https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/ua/uatypes.py
+        /// this should be a member of NodeObject, may be implemented at client class is easier
+        enum PermissionFlag : uint32_t
+        {
+            CurrentRead = 0,
+            CurrentWrite = 1,
+            HistoryRead = 2,
+            HistoryWrite = 3,
+            SemanticChange = 4,
+            StatusWrite = 5,
+            TimestampWrite = 6
+        };
+
+        /// std::filesystem can only check `last_write_time()`
+        /// however, server may not support those time stamps
+        /// consider, merge timestamp into RevisionInfo
+        class TimeInfo
+        {
+        public:
+            TimeInfo(){}; // TODO: set a proper default values for members
+            Poco::DateTime modified;
+
+            std::string timestamp() const
+            {
+                return Poco::DateTimeFormatter::format(modified, TimeInfo::ts_format());
+            }
+            static std::string ts_format()
+            {
+                return "%Y-%m-%dT%H:%M:%S.%F";
+            }
+#if defined(SAL_TIMEINFO_EXTRA)
+            Poco::DateTime created;
+            Poco::DateTime accessed;
+            //
+            static TimeInfo decode(Poco::JSON::Object::Ptr j)
+            {
+                TimeInfo i;
+                int timeZoneDifferential;
+                auto modified_ts = j->getValue<std::string>("modified");
+                i.modified = Poco::DateTimeParser::parse(TimeInfo::ts_format(), modified_ts, timeZoneDifferential);
+                return i;
+            }
+#endif
+            Poco::JSON::Object::Ptr encode()
+            {
+                auto ts = Poco::DateTimeFormatter::format(modified, TimeInfo::ts_format());
+                Poco::JSON::Object::Ptr j = new Poco::JSON::Object();
+                j->set("modified", ts);
+                return j;
+            }
+
+            static TimeInfo decode(std::string modified_ts)
+            {
+                TimeInfo i;
+                int timeZoneDifferential;
+                i.modified = Poco::DateTimeParser::parse(TimeInfo::ts_format(), modified_ts, timeZoneDifferential);
+                return i;
+            }
+        };
+
+        /// integer version, how about timestamp, why not a fixed width int

```

Comment:
Never do this. System time is not reliable. The server persistence layer should track a master tree revision and guarantee a revision number issued to any transaction is unique and sequential.

Reply:
I cant comment on cppsal. This pull request is for the csal interface.

Reply:
got it, just some random thought recorded there